lib/routes/builds.js

```diff
@@ -337,6 +356,43 @@ module.exports = function (app) {
       }).on('error', next);
     });
 
+  app.routes.put('/builds/:pkg/:env?', auth, asynk(async(req, res) => {
+    // Assume no locales are supplied, so just built in en-US, and no artifactory needed
+    const payload = req.body;
+    const builtAssets = payload._attachments || {};
+
+    const models = app.models;
+    const pkg = models.Package.fromPublish(payload);
+
+    // create version and package records in system
+    var version = pkg.version;
+    var name = pkg.name;
+    await manager.addVersionRecord(name, version, payload);
+    await manager.addPackageRecord(pkg);
+
+    // Unpack the base64 string content into a proper directory of code on disk
+    // walk directory to gather files, and publish
+
+    const rootPath = path.join(require('os').tmpdir(), 'untarpkg');
+    const untarPath = path.join(rootPath, `${pkg.name}-${pkg.version}`);
+    const distDirectory = path.join(untarPath, 'package', 'dist');
+    const assetName = `${name}-${version}.tgz`;
```

Comment:
Apologies I was not clear enough – as I said we could assume the "npm publish payload". The additional context / history / software anthropology I neglected to include is that the npm publish payload conforms to the [CouchDB document attachment specification](https://docs.couchdb.org/en/stable/api/document/common.html#attachments) which supports more than tarballs.

For example it should be possible to create a build with multiple plaintext files provided in base64 encoding. e.g.:

``` js
{ _id: 'test-publish03',
  name: 'test-publish03',
  description: 'Just a test',
  'dist-tags': { latest: '1.0.0' },
  versions:
   { '1.0.0':
      { name: 'test-publish03',
        version: '1.0.0',
        description: 'Just a test',
        main: 'index.js',
        scripts: [Object],
        author: '',
        license: 'ISC',
        readme: 'ERROR: No README data found!',
        _id: 'test-publish03@1.0.0',
        _shasum: '6e78734fff0347a9596abb5d3b0360fda6f1a899',
        _from: '.',
        _npmVersion: '3.7.1',
        _nodeVersion: '4.2.2',
        _npmUser: [Object],
        maintainers: [Array],
        dist: [Object] } },
  readme: 'ERROR: No README data found!',
  maintainers: [ { name: 'garbage!!user', email: 'anyvalid@email.com' } ],
  _attachments:
   { 'main.js':
      { content_type: 'application/javascript',
        data: '{{BASE64-ENCODED-FILECONTENTS}}',
        length: 568 } } }
```